console/frontend/src/main/frontend/src/app/services/auth.service.ts

```diff
@@ -1,55 +1,85 @@
-import { Injectable } from '@angular/core';
+import { inject, Injectable } from '@angular/core';
 import { AppService } from '../app.service';
 import { HttpClient } from '@angular/common/http';
-import { Observable } from 'rxjs';
+import { combineLatest, map, Observable, of, tap, shareReplay } from 'rxjs';
+import {
+  Link,
+  LinkName,
+  SecurityItemsService,
+  SecurityItems,
+  SecurityRole,
+  Links,
+} from '../views/security-items/security-items.service';
 
 @Injectable({
   providedIn: 'root',
 })
 export class AuthService {
-  private loggedIn = false;
+  private readonly appService: AppService = inject(AppService);
+  private readonly http: HttpClient = inject(HttpClient);
+  private readonly securityItemsService: SecurityItemsService = inject(SecurityItemsService);
+  private readonly permissionsObservable$: Observable<void>;
 
-  constructor(
-    private http: HttpClient,
-    private appService: AppService,
-  ) {}
+  private loggedIn = false;
+  private allowedRoles: string[] = [];
+  private allowedLinks: Link[] = [];
 
-  /* Currently not being used because servlet handles basic auth */
-  /* login(username: string, password: string): void {
-    if (username != 'anonymous') {
-      this.authToken = this.Base64.encode(`${username}:${password}`);
-      sessionStorage.setItem('authToken', this.authToken);
-    }
-    const location = sessionStorage.getItem('location') || 'status';
-    const absUrl = window.location.href.split('login')[0];
-    window.location.href = absUrl + location;
-    window.location.reload();
-  } */
-  /* loggedin(): void {
-    if (token != null && token != 'null') {
-      if (this.router.url.includes('login'))
-        this.router.navigateByUrl(
-          sessionStorage.getItem('location') || 'status',
-        );
-    } else {
-      if (this.consoleState.init > 0) {
-        if (!this.router.url.includes('login'))
-          sessionStorage.setItem('location', this.router.url || 'status');
-        this.router.navigateByUrl('login');
-      }
-    }
-  } */
+  constructor() {
+    this.permissionsObservable$ = this.initializePermissionsObservable();
+  }
 
   setLoggedIn(username?: string): void {
-    if (username && username != 'anonymous') {
+    if (this.isNotAnonymous(username)) {
       this.loggedIn = true;
     }
   }
 
+  private isNotAnonymous(username?: string): boolean {
+    return !!username && username !== 'anonymous';
+  }
+
+  loadPermissions(): Observable<void> {
+    return this.permissionsObservable$;
+  }
+
+  private initializePermissionsObservable(): Observable<void> {
+    return this.fetchPermissionData().pipe(
+      tap(([securityItems, links]) => this.updatePermissions(securityItems, links)),
+      map(() => {}),
+      shareReplay(1),
+    );
+  }
```

Comment:
It really feels like initializePermissionsObservable should be a promise that calls the fetchPermissionData() and subscribes to it to then call `updatePermissions` with the result instead of having a very weirdly chained together observable which doesnt give back any real value.
I think a promise is a better fit here

Reply:
I think I had it as a promise at first, but that I changed it to only update everything once and to notify everything that's waiting on it. Could you give an example how you'd accieve this with a promise?

Reply:
Thanks for helping me! It is way nicer now.